gst/cgo_exports.go

```diff
@@ -243,3 +243,30 @@ func goPluginInit(plugin *C.GstPlugin, userData C.gpointer) C.gboolean {
 func goGlobalPluginInit(plugin *C.GstPlugin) C.gboolean {
 	return gboolean(globalPluginInit(wrapPlugin(&glib.Object{GObject: glib.ToGObject(unsafe.Pointer(plugin))})))
 }
+
+//export goLogFunction
+func goLogFunction(
+	_ *C.GstDebugCategory,
+	level C.GstDebugLevel,
+	file *C.gchar,
+	function *C.gchar,
+	line C.gint,
+	object *C.GObject,
+	message *C.GstDebugMessage,
+	_ C.gpointer,
+) {
+	if f := customLogFunction; f != nil {
```

Comment:
I believe we also need to take the (RW?)Mutex here. Alternatively, an atomic.Pointer would work here too I believe, or even just documenting that SetLogFunction must be called first thing in main before any goroutine is started...

Reply:
that would be true if we were doing `if customLogFunction != nil`, but setting `f` creates a pointer to the function that won't be changed even if `customLogFunction` is modified during execution

Reply:
The issue here is that without some kind of synchronization/memory barrier, the go memory model (https://go.dev/ref/mem) does not guarantee that some goroutine will see the updated version of customLogFunction if SetLogFunction is called in another goroutine. Since you have a nil check, the consequence would be that the entry doesn't get logged, not a crash, so it may be an acceptable trade off?

Reply:
But in this case, you can also remove the call to logMu on the write side, as locking only on 1 side doesn't really make a difference.